File: src/pokerbots/spam_framework/Card.java
package pokerbots.spam_framework;

import java.util.HashMap;

public class Card {

	private static HashMap<String, Integer> cardIdToNumber;
	
	private String id;
	private int number;
	
	private static void init(){
		String[] suits = { "h", "s", "c", "d" };
		String[] values = { "2", "3", "4", "5", "6", "7", 
				"8", "9", "T", "J", "Q", "K", "A" };
		
		cardIdToNumber = new HashMap<String, Integer>();
		
		
		for (int i = 0; i < values.length; i++){
			for (int j = 0; j < suits.length; j++){
				cardIdToNumber.put(values[i] + suits[j], i*4 + j);
			}
		}
		
	}
	
	public static boolean isCard(String cardId) {
		if (cardIdToNumber == null) {
			init();
		}
		
		return cardIdToNumber.containsKey(cardId);
	}
	
	public Card(String id){
		if (cardIdToNumber == null){
			init();
		}
		
		try {
			if (cardIdToNumber.containsKey(id)) {
				this.id = id;
			} else {
				throw new Exception("Specified card ID is not valid");
			}
		} catch (Exception e) {
			e.printStackTrace();
		}
	}
	
	public String toString(){
		return this.id;
	}
	
	// Return a number between 0 and 51 as a string
	public int toNumber() {
		if(cardIdToNumber == null) {
			init();
		}
		
		return cardIdToNumber.get(this.id);
	}
}

File: spambot/to_server_c++/SPAMFramework/Card.cpp
#include "Card.h"

std::map<std::string, int> Card::card_id_to_number;

Card::Card(std::string id) {
	if (card_id_to_number.size() == 0) {
		Init();
	}

	try {
		// Check if id is in map
		if (IsCard(id)) {
			m_id = id;
		} else {
			throw 10;
		}

	} catch (int e) {
		std::cout << "Specified card ID is not valid" << std::endl;
	}
}

void Card::Init() {
	std::string suits[] = { "h", "s", "c", "d" };
	std::string values[] = { "2", "3", "4", "5", "6", "7", 
				"8", "9", "T", "J", "Q", "K", "A" };

	int num_suits = sizeof(suits) / sizeof(suits[0]);
	int num_values = sizeof(values) / sizeof(values[0]);
	for (int i=0; i < num_values; i++) {
		for (int j=0; j < num_suits; j++) {
			card_id_to_number.insert(std::pair<std::string, int>(values[i]+suits[j], i*4 + j));
		}
	}
}

std::vector<std::string> Card::GetCardList() {
	if (card_id_to_number.size() == 0) {
		Init();
	}

	std::vector<std::string> card_list(52);
	for (std::map<std::string, int>::iterator it = card_id_to_number.begin(); it != card_id_to_number.end(); ++it) {
		card_list.push_back(it->first);
	}

	return card_list;
}

bool Card::IsCard(std::string card_id) {
	if (card_id_to_number.size() == 0) {
		Init();
	}

	return (card_id_to_number.count(card_id) > 0);
}

std::string Card::ToString() {
	return m_id;
}

int Card::ToNumber() {
	if (card_id_to_number.size() == 0) {
		Init();
	}

	return card_id_to_number[m_id];
}
